File: scripts/extract_neo_items.py
#!/usr/bin/env python3
"""Extract item data and images from Neo Scavenger for prototyping."""

import xml.etree.ElementTree as ET
import json
import shutil
import os
import re

NEO_DIR = "/Users/nicholasarehart/Library/Application Support/Steam/steamapps/common/NEO Scavenger"
OUTPUT_DIR = "/Users/nicholasarehart/after-the-end-godot/ui-experiments/src/assets/items"
DATA_OUTPUT = "/Users/nicholasarehart/after-the-end-godot/ui-experiments/src/data/neoItems.json"

# Item groups we care about (from Neo Scavenger)
INTERESTING_GROUPS = {
    5: "container",      # bags, backpacks
    11: "container",     # bottles, containers
    13: "material",      # branches, sticks
    32: "consumable",    # food cans
    33: "consumable",    # food items
    34: "consumable",    # drinks
    1: "weapon",         # melee weapons
    2: "weapon",         # ranged weapons
    3: "ammo",           # ammunition
    6: "clothing",       # clothing
    7: "clothing",       # footwear
    8: "accessory",      # accessories
    9: "tool",           # tools
    10: "medical",       # medical items
}

def parse_size(capacity_str):
    """Parse capacity string like '4x6' into width/height."""
    if not capacity_str or 'x' not in capacity_str:
        return None
    match = re.match(r'(\d+)x(\d+)', capacity_str)
    if match:
        return {"width": int(match.group(1)), "height": int(match.group(2))}
    return None

def get_item_size_from_format(format_id):
    """Estimate item size based on format ID (rough mapping)."""
    # These are approximations based on common Neo Scavenger formats
    sizes = {
        1: {"width": 1, "height": 1},
        2: {"width": 2, "height": 1},
        3: {"width": 1, "height": 1},
        4: {"width": 2, "height": 2},
        5: {"width": 3, "height": 2},
        6: {"width": 4, "height": 2},
        7: {"width": 3, "height": 3},
        8: {"width": 2, "height": 3},
    }
    return sizes.get(format_id, {"width": 1, "height": 1})

def extract_items():
    """Parse itemtypes.xml and extract item data."""
    xml_path = os.path.join(NEO_DIR, "data", "itemtypes.xml")
    tree = ET.parse(xml_path)
    root = tree.getroot()

    items = []
    images_to_copy = set()

    for table in root.findall(".//table[@name='itemtypes']"):
        item = {}
        for col in table.findall("column"):
            item[col.get("name")] = col.text or ""

        # Extract relevant fields
        item_id = item.get("id", "")
        name = item.get("strName", "")
        desc = item.get("strDesc", "")
        group_id = int(item.get("nGroupID", 0))
        weight = float(item.get("fWeight", 0))
        value = float(item.get("fMonetaryValue", 0))
        stack_limit = int(item.get("nStackLimit", 1))
        format_id = int(item.get("nFormatID", 3))
        capacity = item.get("aCapacities", "")
        image_list = item.get("vImageList", "")

        # Get first image from list
        images = [img.strip() for img in image_list.split(",") if img.strip()]
        primary_image = images[0] if images else None

        # Determine item type from group
        item_type = INTERESTING_GROUPS.get(group_id, "misc")

        # Parse container capacity
        grid_size = parse_size(capacity)

        # Estimate item size (grid cells it occupies)
        item_size = get_item_size_from_format(format_id)

        extracted = {
            "id": f"neo_{item_id}",
            "neoId": item_id,
            "name": name,
            "description": desc,
            "type": item_type,
            "weight": weight,
            "value": value,
            "stackLimit": stack_limit,
            "size": item_size,
            "image": primary_image,
            "allImages": images,
        }

        # Add grid size for containers
        if grid_size:
            extracted["gridSize"] = grid_size

        items.append(extracted)

        # Track images to copy
        for img in images:
            if img:
                images_to_copy.add(img)

    return items, images_to_copy

def copy_images(images):
    """Copy item images to output directory."""
    img_dir = os.path.join(NEO_DIR, "img")
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    copied = 0
    for img in images:
        src = os.path.join(img_dir, img)
        dst = os.path.join(OUTPUT_DIR, img)
        if os.path.exists(src):
            shutil.copy2(src, dst)
            copied += 1

    print(f"Copied {copied}/{len(images)} images")

def main():
    print("Extracting items from Neo Scavenger...")
    items, images = extract_items()

    # Create output directories
    os.makedirs(os.path.dirname(DATA_OUTPUT), exist_ok=True)

    # Save item data as JSON
    with open(DATA_OUTPUT, 'w') as f:
        json.dump(items, f, indent=2)
    print(f"Saved {len(items)} items to {DATA_OUTPUT}")

    # Copy images
    print(f"Copying {len(images)} images...")
    copy_images(images)

    # Print summary by type
    by_type = {}
    for item in items:
        t = item["type"]
        by_type[t] = by_type.get(t, 0) + 1

    print("\nItems by type:")
    for t, count in sorted(by_type.items()):
        print(f"  {t}: {count}")

if __name__ == "__main__":
    main()

File: scripts/extractNeoItems.js
#!/usr/bin/env node
// Extract item data and images from Neo Scavenger for prototyping.

import { readFileSync, writeFileSync, mkdirSync, copyFileSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { DOMParser } from '@xmldom/xmldom';

const NEO_DIR =
  '/Users/nicholasarehart/Library/Application Support/Steam/steamapps/common/NEO Scavenger';
const OUTPUT_DIR = '/Users/nicholasarehart/after-the-end-godot/ui-experiments/src/assets/images';
const DATA_OUTPUT =
  '/Users/nicholasarehart/after-the-end-godot/ui-experiments/src/data/neoItems.json';

// Each grid cell in Neo Scavenger is 10 pixels (from GUIInventorySlot.as: nCapacityPixel = 10)
const CELL_SIZE_PIXELS = 10;

const INTERESTING_GROUPS = {
  5: 'container',
  11: 'container',
  13: 'material',
  32: 'consumable',
  33: 'consumable',
  34: 'consumable',
  1: 'weapon',
  2: 'weapon',
  3: 'ammo',
  6: 'clothing',
  7: 'clothing',
  8: 'accessory',
  9: 'tool',
  10: 'medical',
};

const DEFAULT_SIZE = { width: 1, height: 1 };

// Read PNG dimensions from file header
// PNG format: bytes 16-19 = width, bytes 20-23 = height (big endian)
function getPngDimensions(filePath) {
  try {
    const buffer = readFileSync(filePath);
    // Verify PNG signature
    if (buffer[0] !== 0x89 || buffer[1] !== 0x50 || buffer[2] !== 0x4e || buffer[3] !== 0x47) {
      return null;
    }
    const width = buffer.readUInt32BE(16);
    const height = buffer.readUInt32BE(20);
    return { width, height };
  } catch {
    return null;
  }
}

// Calculate grid size from image dimensions
function calculateGridSize(imagePath) {
  const dimensions = getPngDimensions(imagePath);
  if (!dimensions) return DEFAULT_SIZE;

  return {
    width: Math.round(dimensions.width / CELL_SIZE_PIXELS),
    height: Math.round(dimensions.height / CELL_SIZE_PIXELS),
  };
}

const str = (val) => val ?? '';
const int = (val, fallback) => parseInt(val ?? fallback, 10);
const float = (val) => parseFloat(val ?? '0');
const toCamelCase = (filename) => filename.charAt(0).toLowerCase() + filename.slice(1);

function parseSize(capacityStr) {
  const match = capacityStr?.match(/(\d+)x(\d+)/);
  return match ? { width: int(match[1], 0), height: int(match[2], 0) } : null;
}

function parseImages(imageList) {
  return str(imageList)
    .split(',')
    .map((img) => img.trim())
    .filter(Boolean);
}

// Parse vImageUsage to get inventory image index
// Format: "ground_idx,inventory_idx,..." - we need the second value (inventory)
function getInventoryImageIndex(imageUsage) {
  const parts = str(imageUsage)
    .split(',')
    .map((n) => parseInt(n, 10));
  return parts.length > 1 ? parts[1] : 0;
}

function parseTableToObject(table) {
  const item = {};
  const columns = table.getElementsByTagName('column');
  for (let j = 0; j < columns.length; j++) {
    const col = columns[j];
    item[col.getAttribute('name')] = col.textContent;
  }
  return item;
}

function extractItemData(item, imgDir) {
  const originalImages = parseImages(item.vImageList);
  const camelImages = originalImages.map(toCamelCase);
  const groupId = int(item.nGroupID, 0);
  const gridSize = parseSize(item.aCapacities);

  // Get the inventory image (vImageUsage specifies which image is used in inventory)
  const inventoryImageIndex = getInventoryImageIndex(item.vImageUsage);
  const inventoryImage = originalImages[inventoryImageIndex] ?? originalImages[0];
  const imagePath = inventoryImage ? join(imgDir, inventoryImage) : null;
  const size = imagePath ? calculateGridSize(imagePath) : DEFAULT_SIZE;

  const extracted = {
    id: `neo_${str(item.id)}`,
    neoId: str(item.id),
    name: str(item.strName),
    description: str(item.strDesc),
    type: INTERESTING_GROUPS[groupId] ?? 'misc',
    weight: float(item.fWeight),
    value: float(item.fMonetaryValue),
    stackLimit: int(item.nStackLimit, 1),
    size,
    image: camelImages[0] ?? null,
    allImages: camelImages,
    ...(gridSize && { gridSize }),
  };

  return { extracted, originalImages };
}

function extractItems() {
  const xmlPath = join(NEO_DIR, 'data', 'itemtypes.xml');
  const imgDir = join(NEO_DIR, 'img');
  const doc = new DOMParser().parseFromString(readFileSync(xmlPath, 'utf-8'), 'text/xml');

  const items = [];
  const imagesToCopy = new Map();
  const tables = doc.getElementsByTagName('table');

  for (let i = 0; i < tables.length; i++) {
    if (tables[i].getAttribute('name') !== 'itemtypes') continue;

    const item = parseTableToObject(tables[i]);
    const { extracted, originalImages } = extractItemData(item, imgDir);
    items.push(extracted);
    originalImages.forEach((img) => imagesToCopy.set(img, toCamelCase(img)));
  }

  return { items, imagesToCopy };
}

function copyImages(imageMap) {
  const imgDir = join(NEO_DIR, 'img');
  mkdirSync(OUTPUT_DIR, { recursive: true });

  let copied = 0;
  for (const [original, camel] of imageMap) {
    const src = join(imgDir, original);
    if (existsSync(src)) {
      copyFileSync(src, join(OUTPUT_DIR, camel));
      copied++;
    }
  }
  console.log(`Copied ${copied}/${imageMap.size} images`);
}

function printSummary(items) {
  const byType = {};
  for (const item of items) {
    byType[item.type] = (byType[item.type] || 0) + 1;
  }
  console.log('\nItems by type:');
  for (const [t, count] of Object.entries(byType).sort()) {
    console.log(`  ${t}: ${count}`);
  }
}

function main() {
  console.log('Extracting items from Neo Scavenger...');
  const { items, imagesToCopy } = extractItems();

  mkdirSync(dirname(DATA_OUTPUT), { recursive: true });
  writeFileSync(DATA_OUTPUT, JSON.stringify(items, null, 2));
  console.log(`Saved ${items.length} items to ${DATA_OUTPUT}`);

  console.log(`Copying ${imagesToCopy.size} images...`);
  copyImages(imagesToCopy);
  printSummary(items);
}

main();
